Type/PasswordHashBase.pm

```diff
@@ -0,0 +1,84 @@
+# Copyright (c) 2023 bivio Software, Inc.  All rights reserved.
+package Bivio::Type::PasswordHashBase;
+use strict;
+use Bivio::Base 'Type.SyntacticString';
+
+my($_IDI) = __PACKAGE__->instance_data_index;
+my(@_SALT_CHARS) = (
+    'a'..'z',
+    'A'..'Z',
+    '0'..'9',
+);
+my($_SALT_INDEX_MAX) = int(@_SALT_CHARS) - 1;
+
+sub ID {
+    b_die('abstract method');
+}
+
+sub REGEX {
+    b_die('abstract method');
+}
+
+sub SALT_LENGTH {
+    b_die('abstract method');
+}
+
+sub as_literal {
+    my($self) = @_;
+    return $self->internal_format_literal($self->get_salt, $self->get_hash);
+}
+
+sub compare {
+    my($self, $clear_text) = @_;
+    return $self->as_literal cmp $self->to_literal($clear_text, $self->get_salt);
+}
+
+sub get_id {
+    return shift->[$_IDI][0];
```

Comment:
I don't generally like this style. It's better to have a hash rather than an array so there is no implicit coupling on the ordering. The coupling is always going to be implicit in internal_to_parts and internal_post_from_literal. The "to parts" is ordered implicitly and cascaded subclasses. Just noting that it might be better to not have "internal_to_parts" this way and instead return a hash from internal_to_parts.

Probably doesn't matter, just my personal opinion.